contracts/FIFOMarket.sol

```diff
@@ -227,19 +252,38 @@ contract FIFOMarket is
       batchedAmounts,
       encodedCertificateAmount
     );
+    address[] memory feeRecipients = new address[](highestFeeIndex + 1);
+    uint256[] memory fees = new uint256[](highestFeeIndex + 1);
+    bytes[] memory data = new bytes[](highestFeeIndex + 1); // todo overload batchSend to allow no data params
+    bool[] memory requireReceptionAck = new bool[](highestFeeIndex + 1); // todo overload batchSend to allow no requireReceptionAck params
+    fees[noriFee.index] = noriFee.fee;
+    feeRecipients[noriFee.index] = noriFee.account;
+    data[noriFee.index] = "0x";
+    requireReceptionAck[noriFee.index] = false;
     for (uint256 i = 0; i < batchedIds.length; i++) {
       if (
         batchedAmounts[i] == _removal.balanceOf(address(this), batchedIds[i])
       ) {
         _queueHeadIndex++;
       }
-      totalSupply -= batchedAmounts[i];
-      uint256 noriFee = (batchedAmounts[i] / 100) * _noriFee;
-      uint256 supplierFee = batchedAmounts[i];
-      _bridgedPolygonNori.transfer(_noriFeeWallet, noriFee);
-      _bridgedPolygonNori.transfer(suppliers[i], supplierFee);
+      totalSupply -= batchedAmounts[i]; // todo test gas optimization with caching of these vars
+    }
+    for (uint256 i = 0; i < highestFeeIndex; i++) {
```

Comment:
this loop iterates over the number of *unique* suppliers whose removals went into fulfilling an order (because there's only one fee for any given supplier no matter how many removals they contributed).  however the `suppliers` array that you're indexing with `i` here on line 272 may contain duplicate suppliers because a supplier address is pushed onto that array for every removal that comprises the order event if multiple come from the same supplier.  I think this needs to be fixed so that when you're populating these `fees` and `feeRecipients` arrays for the batched send you're ONLY looping over the fee structs for the order (which would involve making them indexable by index somehow).  or you create an enumerable set of *unique* suppliers for the order and iterate over that, grabbing the fee struct for each address.

Reply:
good catch-- I like the enumerable set approach as it might end up cleaning up things a little bit